src/util/psbt/map/global.rs

```diff
@@ -0,0 +1,178 @@
+// Rust Bitcoin Library
+// Written by
+//   The Rust Bitcoin developers
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+use std::collections::HashMap;
+use std::io::Cursor;
+
+use blockdata::transaction::Transaction;
+use consensus::encode::{self, Encodable, Decodable, Decoder};
+use util::psbt::map::Map;
+use util::psbt::raw;
+use util::psbt;
+use util::psbt::Error;
+
+/// A key-value map for global data.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Global {
+    /// The unsigned transaction, scriptSigs and witnesses for each input must be
+    /// empty.
+    pub unsigned_tx: Transaction,
+    /// Unknown global key-value pairs.
+    pub unknown: HashMap<raw::Key, Vec<u8>>,
+}
+
+impl Global {
+    /// Create a Global from an unsigned transaction, error if not unsigned
+    pub fn from_unsigned_tx(tx: Transaction) -> Result<Self, psbt::Error> {
+        for txin in &tx.input {
+            if !txin.script_sig.is_empty() {
+                return Err(Error::UnsignedTxHasScriptSigs);
+            }
+
+            if !txin.witness.is_empty() {
+                return Err(Error::UnsignedTxHasScriptWitnesses);
+            }
+        }
+
+        Ok(Global {
+            unsigned_tx: tx,
+            unknown: Default::default(),
+        })
+    }
+}
+
+impl Map for Global {
+    fn insert_pair(&mut self, pair: raw::Pair) -> Result<(), encode::Error> {
+        let raw::Pair {
+            key: raw_key,
+            value: raw_value,
+        } = pair;
+
+        match raw_key.type_value {
+            0u8 => {
+                return Err(Error::DuplicateKey(raw_key).into());
+            }
+            _ => {
+                if self.unknown.contains_key(&raw_key) {
+                    return Err(Error::DuplicateKey(raw_key).into());
+                } else {
+                    self.unknown.insert(raw_key, raw_value);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_pairs(&self) -> Result<Vec<raw::Pair>, encode::Error> {
+        let mut rv: Vec<raw::Pair> = Default::default();
+
+        rv.push(raw::Pair {
+            key: raw::Key {
+                type_value: 0u8,
+                key: vec![],
+            },
+            value: {
+                let mut ret = Vec::new();
+                self.unsigned_tx.version.consensus_encode(&mut ret)?;
+                self.unsigned_tx.input.consensus_encode(&mut ret)?;
+                self.unsigned_tx.output.consensus_encode(&mut ret)?;
+                self.unsigned_tx.lock_time.consensus_encode(&mut ret)?;
+                ret
```

Comment:
Only nit: Add a comment explaining why you're doing the manual serialization here. New readers will be confused otherwise.